### PR TITLE
karmada-search: optimize backend store lock

### PR DESCRIPTION
**Pull request by calvin0327**
Signed-off-by: calvin <wen.chen@daocloud.io>

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
In most cases it's reading, a small optimization to separate reading and writing. 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
please correct me if not.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by RainbowMango** [comment]
Is it a bug? (I see you tagged the PR with `kind/bug`)

**Comment by RainbowMango** [comment]
I guess it is more like a cleanup or improvement. (We don't have a `kind/improvement` label though)

**Comment by calvin0327** [comment]
> I guess it is more like a cleanup or improvement. (We don't have a `kind/improvement` label though)

👌

File: pkg/search/backendstore/store.go [hunk 1]
```diff
@@ -62,12 +62,19 @@ func DeleteBackend(cluster string) {
...
 	backendLock.Lock()
 	defer backendLock.Unlock()
-	bs, ok := backends[cluster]
-	if !ok {
-		klog.Errorf("cannot find backend store %s, use defult backend", cluster)
-		return NewDefaultBackend(cluster)
-	}
-	return bs
+
+	klog.Errorf("cannot find backend store %s, use defult backend", cluster)
```
Comment (ikaven1024) [inline pkg/search/backendstore/store.go:78]: is it necessary lock for `NewDefaultBackend`?

File: pkg/search/backendstore/store.go [hunk 2]
```diff
@@ -62,12 +62,19 @@ func DeleteBackend(cluster string) {
...
+		backendLock.RLock()
+		defer backendLock.RUnlock()
+
+		bs, ok := backends[cluster]
+		return bs, ok
+	}(); ok {
+		return bs
+	}
+
 	backendLock.Lock()
```
Comment (ikaven1024) [inline pkg/search/backendstore/store.go:75]: is it necessary lock for NewDefaultBackend?

Comment (liys87x) [reply pkg/search/backendstore/store.go:75]: unnecessary

Comment (calvin0327) [reply pkg/search/backendstore/store.go:75]: If there are other goroutines writing to the backends map, `GetBackend` should be locked.  but is it necessary to use read/write locks? please corrent me if not

File: pkg/search/backendstore/store.go (hunk 1 above)
Comment (chaunceyjiang) [inline pkg/search/backendstore/store.go:78]: klog.Errorf -> klog.Warning ?

Comment (liys87x) [reply pkg/search/backendstore/store.go:78]: This path does not meet user expectations, So the error log was selected

**Comment by liys87x** [comment]
The `default` backend store does nothing, but the `singleClusterInformerManagerImpl.ForResource` method must provide a handler.
In the future, we will remove the `default` backen store if we can.

**Comment by liys87x** [comment]
This is a low frequency operation and returning the default backend does not meet user expectations. We will not deal with this for the time being, and we will adjust it together in the future. 
What do you think? @calvin0327

**Comment by calvin0327** [comment]
@liys87x okay.

**Pull request closed without merging**